src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,876 @@
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var typeName = WellKnownTypeNames.SystemOperatingSystem;
+
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(typeName + "Helper", out var operatingSystemType))
+                {
+                    // TODO: remove 'typeName + "Helper"' after tests able to consume the real new APIs
+                    operatingSystemType = context.Compilation.GetOrCreateTypeByMetadataName(typeName);
+                }
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesOSPlatform, out var osPlatformType) ||
+                    !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesRuntimeInformation, out var runtimeInformationType))
+                {
+                    return;
+                }
+
+                var runtimeIsOSPlatformMethod = runtimeInformationType.GetMembers().OfType<IMethodSymbol>().Where(m =>
+                    IsOSPlatform == m.Name &&
+                    m.IsStatic &&
+                    m.ReturnType.SpecialType == SpecialType.System_Boolean &&
+                    m.Parameters.Length == 1 &&
+                    m.Parameters[0].Type.Equals(osPlatformType)).FirstOrDefault();
+
+                var guardMethods = GetRuntimePlatformGuardMethods(runtimeIsOSPlatformMethod, operatingSystemType!);
+
+                context.RegisterOperationBlockStartAction(context => AnalyzeOperationBlock(context, guardMethods, osPlatformType));
+            });
+
+            static ImmutableArray<IMethodSymbol> GetRuntimePlatformGuardMethods(IMethodSymbol runtimeIsOSPlatformMethod, INamedTypeSymbol operatingSystemType)
+            {
+                return operatingSystemType.GetMembers().OfType<IMethodSymbol>().Where(m =>
+                    s_platformCheckMethodNames.Contains(m.Name) &&
+                    m.IsStatic &&
+                    m.ReturnType.SpecialType == SpecialType.System_Boolean).ToImmutableArray().
+                    Add(runtimeIsOSPlatformMethod);
+            }
+        }
+
+        private void AnalyzeOperationBlock(OperationBlockStartAnalysisContext context, ImmutableArray<IMethodSymbol> guardMethods, INamedTypeSymbol osPlatformType)
+        {
+            var platformSpecificOperations = PooledConcurrentDictionary<IOperation, SmallDictionary<string, PlatformAttributes>>.GetInstance();
+            var platformSpecificMembers = PooledConcurrentDictionary<ISymbol, SmallDictionary<string, PlatformAttributes>>.GetInstance();
```

Comment:
We should experiment this cache having the lifetime of whole compilation instead of just each operation block. I will try it out and post the numbers tomorrow.